OAK-9790 - Implement parallel indexing for speeding up oak run indexing command

amit-jain opened the pull request:
- Parallel indexing enabled by splitting the FlatFileStore taking into account aggregation and then running multiple threads to do the indexing
- LZ4 compression enabled

Work over https://github.com/apache/jackrabbit-oak/pull/587

Commit:
OAK-9790 - Implement parallel indexing for speeding up oak run indexing command

- Parallel indexing enabled by splitting the FlatFileStore taking into account aggregation and then running multiple threads to do the indexing
- LZ4 compression enabled

amit-jain commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:
```diff
@@ -265,6 +271,55 @@ public void reindex() throws CommitFailedException, IOException {
+            }
+            log.info("All {} indexing jobs are done", storeList.size());
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Failure getting indexing job result", e);
```
@fabriziofortino You had a comment in the original PR which I forgot to address. It might be   better to throw in case of exceptions as it affects correctness. @thomasmueller wdyt?

fabriziofortino replied:
I would propagate the exception and fail the indexing job.

nit0906 commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java:
```diff
@@ -0,0 +1,255 @@
+ * This class is being used when {@value IndexerConfiguration#PROP_OAK_INDEXER_PARALLEL_INDEX} is set to true.
+ * It will split a flat file safely by checking the index definitions. An entry is considered safe to split if only
+ * none of the parent directories contains nodes in indexRule and aggregate fields of the provided index definitions.
+ */
```
An example would be beneficial here.

amit-jain replied:
Examples are available as part of the tests, not sure it makes sense to list them here.

nit0906 commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java:
```diff
@@ -0,0 +1,255 @@
+        return splitFlatFiles;
+    }
+
+    private void updateNodeTypeStack(Stack<String> parentNodeTypeNames, String line) {
```
There are a few util methods in this class like this one - I wonder if it might make sense to make these static to be able to easily write test cases for these ?

amit-jain replied:
Possibly we can, but these steps should mostly be covered in the FlatFileSplitterTest. Do you think we are missing tests here.

gnaresh19 commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java:
```diff
@@ -151,6 +153,7 @@ private BulkProcessor initBulkProcessor() {
         return BulkProcessor.builder(requestConsumer(),
                 new OakBulkProcessorListener(), this.indexName + "-bulk-processor")
                 .setBulkActions(indexDefinition.bulkActions)
+                .setConcurrentRequests(BULK_PROCESSOR_CONCURRENCY)
```
@amit-jain : Can the concurrentRequests property be moved to index definition, similar to other properties like bulkSize, which would enable the caller to configure appropriate concurrentRequests in elastic index definition.

amit-jain replied:
This config does not seem suitable to be included in the IndexDefinition as it is affecting the processing in general and not per index. @nit0906 wdyt?

nit0906 replied:
@amit-jain  - agreed.

the bulkProcessor instance is actually linked to index definition and all other configurations like bulk size threshold , flush interval etc are also defined/configured at index definition level. So if we want to keep configurations in the same line - then it would seem the concurrency can also go in the index definition. 

But this is a different case where we would want to distinguish this on the basis of whether indexing is running from oak-run or otherwise - so configuring this in index definition won't be possible (since the index def would remain the same in both cases but we would want the concurrency to be different - so a system property is the only solution here.)

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/IndexWriterUtils.java:
```diff
@@ -33,12 +35,13 @@
-import static org.apache.jackrabbit.oak.plugins.index.lucene.LuceneIndexConstants.VERSION;
-
 public class IndexWriterUtils {
+    private static final int INDEX_WRITER_MAX_MERGE = 8;
```
this should be configurable imo

amit-jain replied:
Yes, makes sense.

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/IndexWriterUtils.java:
```diff
@@ -70,6 +73,10 @@ public static IndexWriterConfig getIndexWriterConfig(LuceneIndexDefinition defin
             IndexWriterConfig config = new IndexWriterConfig(VERSION, analyzer);
             if (remoteDir) {
```
one reason why remoteDir could be true is because copyOnWrite is disabled (which usually isn't the case).

So effectively the new code change here (the else block) will impact normal indexing as well (not just the out of the band indexing from oak-run), from what I can tell - was that taken into account ?

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/IndexWriterUtils.java:
```diff
@@ -70,6 +73,10 @@ public static IndexWriterConfig getIndexWriterConfig(LuceneIndexDefinition defin
             IndexWriterConfig config = new IndexWriterConfig(VERSION, analyzer);
             if (remoteDir) {
                 config.setMergeScheduler(new SerialMergeScheduler());
+            } else {
```
one reason why remoteDir could be true is because copyOnWrite is disabled (which usually isn't the case), which would imply we would most probably run into the else block here even during normal indexing.

So effectively the new code change here (the else block) will impact normal indexing as well (not just the out of the band indexing from oak-run), from what I can tell - was that taken into account ?

amit-jain replied:
I am not sure. Can you suggest what change would be advisable here to account for that. The default is 1 thread so essentially the concurrent will default to serial ofcourse if the property is not overriden

nit0906 replied:
> The default is 1 thread so essentially the concurrent will default to serial ofcourse if the property is not overriden

How's the default 1 ? The max threads are hardcoded to 8 as of now - or did I miss something ?

amit-jain replied:
As you suggested the properties should be configurable and if that's the case then the default should be 1. But question is what can be done if in future this default changes?

nit0906 replied:
ah - ok. I think it would be difficult to determine here whether this is being called from out-of-band indexing or regular async indexing. So I think the best we can do is to set the default to 1 and add a comment that the default value should be changed only after thorough analysis.

Commit:
OAK-9790 - Implement parallel indexing for speeding up oak run indexing command

Changes from review
- Added a unit test for FlatFileSplitter & made some internal methods static
- make some properties configurable

Commit:
Merge branch 'trunk' into OAK-9790

# Conflicts:
#	oak-lucene/pom.xml

Commit:
OAK-9790 - Throw exception and not just log in case of task result failure

Commit:
OAK-9790 - Minor cleanup

Commit:
OAK-9790 - Embedding org.lz4:lz4-java in oak-commons as needed in many modules and removing as test dependency

Commit:
OAK-9790 - IT- Slightly reducing the number of nodes created

amit-jain commented:
Alternate [changes](https://github.com/amit-jain/jackrabbit-oak/commit/9d0c53b62e8ff8a10438052796522afc4d6654c0) to remove dependency on the LZ4 for other modules and only introduce it for oak-run-commons/oak-run

thomasmueller commented on oak-commons/pom.xml:
```diff
@@ -58,6 +58,9 @@
             </Export-Package>
+            <Embed-Dependency>
+              lz4-java;scope=compile|runtime
+            </Embed-Dependency>
```
For this change, I think we should ask others in the Oak team... @reschke , @mreutegg , @anchela

amit-jain replied:
> This is a big change, and I think we need to be careful to not break anything. I think we will need more reviewers.

This is a PR to incorporate changes from review of https://github.com/apache/jackrabbit-oak/pull/587. So, its been reviewed by many reviewers already

thomasmueller reviewed (commented):
This is a big change, and I think we need to be careful to not break anything. I think we will need more reviewers.

joerghoh commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/progress/IndexingProgressReporter.java:
```diff
@@ -229,10 +231,12 @@ public IndexUpdateState(String indexPath, boolean reindex, long estimatedCount)
-                watch.reset().start();
+            long count = updateCount.incrementAndGet();
+            if (count % 10000 == 0) {
+                synchronized (this) {
```
Is there a need to synchronize this?

amit-jain replied:
The Stopwatch class is not thread-safe so, potentially could result in instability otherwise

Commit:
OAK-9790 - Removing support for LZ4 compression

amit-jain commented:
LZ4 support moved to issue OAK-9968 & commit https://github.com/amit-jain/jackrabbit-oak/commit/e8b4a0617fc857f281d6dec967cd48f7e5a16450
